streams/src/main/java/org/apache/kafka/streams/state/internals/BufferValue.java

```diff
@@ -16,83 +16,162 @@
+        } else {
+            oldValue = new byte[oldValueLength];
+            buffer.get(oldValue);
+        }
+
+        final byte[] newValue = extractValue(buffer);
+
+        return new BufferValue(priorValue, oldValue, newValue, context);
+    }
 
-        final int priorValueLength = buffer.getInt();
-        if (priorValueLength == -1) {
-            return new BufferValue(record, null);
+    private static byte[] extractValue(final ByteBuffer buffer) {
+        final int valueLength = buffer.getInt();
+        if (valueLength == NULL_VALUE_SENTINEL) {
+            return null;
         } else {
-            final byte[] priorValue = new byte[priorValueLength];
-            buffer.get(priorValue);
-            return new BufferValue(record, priorValue);
+            final byte[] value = new byte[valueLength];
+            buffer.get(value);
+            return value;
         }
     }
 
     ByteBuffer serialize(final int endPadding) {
 
-        final int sizeOfPriorValueLength = Integer.BYTES;
-        final int sizeOfPriorValue = priorValue == null ? 0 : priorValue.length;
+        final int sizeOfValueLength = Integer.BYTES;
+
+        final int sizeOfPriorValue = priorValue == null || priorValue == oldValue ? 0 : priorValue.length;
+        final int sizeOfOldValue = oldValue == null ? 0 : oldValue.length;
+        final int sizeOfNewValue = newValue == null ? 0 : newValue.length;
+
+        final byte[] serializedContext = recordContext.serialize();
+
+        final ByteBuffer buffer = ByteBuffer.allocate(
+            serializedContext.length
+                + sizeOfValueLength + sizeOfPriorValue
+                + sizeOfValueLength + sizeOfOldValue
+                + sizeOfValueLength + sizeOfNewValue
+                + endPadding
+        );
 
-        final ByteBuffer buffer = record.serialize(sizeOfPriorValueLength + sizeOfPriorValue + endPadding);
+        buffer.put(serializedContext);
 
-        if (priorValue == null) {
-            buffer.putInt(-1);
+        addValue(buffer, priorValue);
+
+        if (oldValue == null) {
+            buffer.putInt(NULL_VALUE_SENTINEL);
+        } else if (priorValue == oldValue) {
+            buffer.putInt(OLD_PREV_DUPLICATE_VALUE_SENTINEL);
         } else {
-            buffer.putInt(priorValue.length);
-            buffer.put(priorValue);
+            buffer.putInt(sizeOfOldValue);
+            buffer.put(oldValue);
         }
 
+        addValue(buffer, newValue);
+
         return buffer;
     }
 
+    private static void addValue(final ByteBuffer buffer, final byte[] newValue) {
+        if (newValue == null) {
+            buffer.putInt(NULL_VALUE_SENTINEL);
+        } else {
+            buffer.putInt(newValue.length);
+            buffer.put(newValue);
+        }
+    }
+
     long sizeBytes() {
-        return (priorValue == null ? 0 : priorValue.length) + record.sizeBytes();
+        return (priorValue == null || priorValue == oldValue ? 0 : priorValue.length)
```

Comment:
@guozhangwang , regarding https://github.com/vvcephei/kafka/pull/1#discussion_r289221021

> This seems not correct: we may not encode the old value at all right?

Yeah, I got this backwards. I was initially skipping to encode the _prior_ value, but now I'm skipping to encode the _old_ value (because the deserialization works out better that way), but I didn't update these cases.

Reply:
I've corrected it now.